internal/exposer/exposer.go

```diff
@@ -0,0 +1,488 @@
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/siderolabs/kube-service-exposer/internal/version"
+)
+
+var _ reconcile.Reconciler = &Exposer{}
+
+// New creates a new Exposer.
+func New(annotationKey string, bindCIDRs []string, logger *zap.Logger) (*Exposer, error) {
+	errs := validation.ValidateAnnotations(map[string]string{annotationKey: "65535"}, field.NewPath("metadata", "annotations"))
+	if len(errs) > 0 {
+		return nil, fmt.Errorf("invalid annotation key: %w", errs.ToAggregate())
+	}
+
+	if logger == nil {
+		logger = zap.NewNop()
+	}
+
+	conf, err := config.GetConfig()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get config: %w", err)
+	}
+
+	mgr, err := manager.New(conf, manager.Options{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to create manager: %w", err)
+	}
+
+	bindCIDRPrefixes := make([]netip.Prefix, 0, len(bindCIDRs))
+
+	for _, bindCIDR := range bindCIDRs {
+		prefix, err := netip.ParsePrefix(bindCIDR)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse bindCIDR: %w", err)
+		}
+
+		bindCIDRPrefixes = append(bindCIDRPrefixes, prefix)
+	}
+
+	return &Exposer{
+		AnnotationKey:     annotationKey,
+		BindCIDRPrefixes:  bindCIDRPrefixes,
+		manager:           mgr,
+		logger:            logger,
+		hostPortToMapping: map[int]*portMapping{},
+		svcToLB:           map[string]*loadBalancer{},
+	}, nil
+}
+
+type portMapping struct {
+	hostIPSet map[string]struct{}
+	svcName   string
+	svcPort   int
+}
+
+type loadBalancer struct {
+	lb       loadbalancer.TCP
+	hostPort int
+}
+
+// Exposer is a controller that exposes the given services on the given host interfaces.
+type Exposer struct {
+	manager           manager.Manager
+	hostPortToMapping map[int]*portMapping
+	svcToLB           map[string]*loadBalancer
+	logger            *zap.Logger
+	ipSet             map[string]struct{}
+	AnnotationKey     string
+	BindCIDRPrefixes  []netip.Prefix
+	mappingLock       sync.Mutex
+	ipSetLock         sync.Mutex
+}
+
+// Run runs the Exposer.
+func (e *Exposer) Run(ctx context.Context) error {
+	e.logger.Info(
+		"starting Exposer",
+		zap.String("annotationKey", e.AnnotationKey),
+		zap.Strings("bindCIDRs", slices.Map(e.BindCIDRPrefixes, func(prefix netip.Prefix) string {
+			return prefix.String()
+		})),
+	)
+
+	if len(e.BindCIDRPrefixes) == 0 {
+		e.logger.Info("bindCIDRs are empty, mappings will listen on all interfaces")
+	}
+
+	ctrller, err := controller.New(version.Name+"-controller", e.manager,
+		controller.Options{
+			Reconciler:         e,
+			NeedLeaderElection: pointer.Bool(false),
+		})
+	if err != nil {
+		return fmt.Errorf("failed to create controller: %w", err)
+	}
+
+	if err = ctrller.Watch(source.Kind(e.manager.GetCache(), &corev1.Service{}), &handler.EnqueueRequestForObject{}); err != nil {
+		return fmt.Errorf("failed to watch Services: %w", err)
+	}
+
+	var eg errgroup.Group
```

Comment:
```suggestion
	eg, ctx := errgroup.WithContext(ctx)
```

This important because otherwise if either `e.manager.Start` or `runIPChangeListener` return an error, the context will not be canceled and app will enter invalid state (I suppose?).

Using `errgroup.Group` is not that great either, because if any `func() error` passed to `eg.Go` return without an error it will not cancel the returned context, so the other closure will continue its execution. You can look in here how it was solved (here)[https://github.com/siderolabs/omni/blob/main/internal/pkg/errgroup/errgroup.go#L4-L3]

Reply:
Good point. I changed it to be a simpler version of our internal solution, to avoid creating the wrapper. I think this would be enough.